src/libmscore/scorefile.cpp

```diff
@@ -449,16 +449,7 @@ bool MasterScore::saveFile(bool generateBackup)
             dir.mkdir(backupSubdirString);
 #ifdef Q_OS_WIN
             const QString backupDirNativePath = QDir::toNativeSeparators(backupDirString);
-#if (defined (_MSCVER) || defined (_MSC_VER))
-   #if (defined (UNICODE))
-            SetFileAttributes((LPCTSTR)backupDirNativePath.unicode(), FILE_ATTRIBUTE_HIDDEN);
-   #else
-            // Use byte-based Windows function
-            SetFileAttributes((LPCTSTR)backupDirNativePath.toLocal8Bit(), FILE_ATTRIBUTE_HIDDEN);
-   #endif
-#else
-            SetFileAttributes((LPCTSTR)backupDirNativePath.toLocal8Bit(), FILE_ATTRIBUTE_HIDDEN);
-#endif
+            SetFileAttributes(reinterpret_cast<LPCTSTR>(backupDirNativePath.unicode()->unicode()), FILE_ATTRIBUTE_HIDDEN);
```

Comment:
@igorkorsukov, `reinterpret_cast` was necessary to stop Qt Creator complaining.

It still seems a bit wierd to do [`QString::unicode()`](https://doc.qt.io/qt-5/qstring.html#unicode), which returns `const QChar *` (i.e. a pointer to a sequence of QChars), and then call [`QChar::unicode()`](https://doc.qt.io/qt-5/qchar.html#unicode) on just the first `QChar` in the sequence, or am I missing something here? Surely we need to end up with a pointer, but doesn't `QChar::unicode()` just return a codepoint?

Reply:
Yet another sign that this `->unicode()` isn't needed, and now even does harm.

Reply:
As I said, the cast `QChar*` to `LPCTSTR` is not very correct, it only works thanks to the current implementation of  `QChar*`, it works because `sizeof( QChar) == sizeof(uint16_t)`. If the internal implementation of `QChar` changes, some field will be added, for example `QChar` will be like this:
```
class QChar {
chort* data;
bool someInternalField;
}
```
then everything will break.
It is unlikely that there will be a change in the implementation of `QChar`, that is, in this case, we are talking about the formal correctness of the work.

Reply:
So let's cross that bridge whe we get there. Currently the code just works (and did all the way up to now), and does not need that `->unicode()`. I don't think Qt will change their implementation, I also don't think we're going to move do any Qt version higher than 5.15.2, as none of those are a) LTS and b) free to use anymore.

Whether we change from a C-style cast to a C++ style one is a matter of taste here

Reply:
> I also don't think we're going to move do any Qt version higher than 5.15.2, as none of those are a) LTS and b) free to use anymore.

But we are creating an application for long-term use and development, in a few years there will be Qt 6 and Qt 7 ... :)

Reply:
As said, let's rather cross that brigdge when we get there.

Reply:
let's cross that bridge

Reply:
```
reinterpret_cast<LPCTSTR>(backupDirNativePath.unicode()->unicode())
```
Is not correct :(
Correct
```
reinterpret_cast<LPCTSTR>(&backupDirNativePath.unicode()->unicode())
```

or I think it better, just 

```
backupDirNativePath.utf16()
```

Reply:
No cast needed? Then I like it better ;-)

Reply:
Nope, needs a cast

Reply:
I thought again
```
reinterpret_cast<LPCTSTR>(&backupDirNativePath.unicode()->unicode())
```
and 
```
&backupDirNativePath.unicode()->unicode()
```
not correct too, sorry for misleading

So
```
reinterpret_cast<LPCTSTR>(backupDirNativePath.unicode())
```
not correct, but work by the internal implementation 

```
backupDirNativePath.utf16()
``
correct and work

Reply:
No, it doesn't even compile:

    libmscore\scorefile.cpp:450:62: error: invalid conversion from 'const ushort*' {aka 'const short unsigned int*'} to 'LPCWSTR' {aka 'const wchar_t*'} [-fpermissive]
                   SetFileAttributes(backupDirNativePath.utf16(), FILE_ATTRIBUTE_HIDDEN);

Reply:
Eh, it means we need a cast, but it's still the most correct option.
```
reinterpret_cast<LPCTSTR>(backupDirNativePath.utf16())
```
The specification is guaranteed that: 
ushort - 16 bit
wchar_t - 16 bit for Windows 

https://en.cppreference.com/w/cpp/language/types

Reply:
That does compile at least ;-), but isn't utf16 also unicode? And a method called `unicode()` pretty well matches the other part of the PR, `-DUNICODE` and `-D_UNICODE`

Reply:
> And a method called unicode() pretty well matches the other part of the PR, -DUNICODE and -D_UNICODE

Nope, this is just an embarrassment, because the unicode() method just returns a pointer to one of the Qt classes :)

Reply:
> but isn't utf16 also unicode? 

yes! 
"Unicode can be implemented by different character encodings. The Unicode standard defines Unicode Transformation Formats (UTF) UTF-8, UTF-16, and UTF-32, and several other encodings. The most commonly used encodings are UTF-8, UTF-16, and UCS-2 (a precursor of UTF-16 without full support for Unicode); GB18030 is standardized in China and implements Unicode fully, while not an official Unicode standard."
https://en.wikipedia.org/wiki/Unicode

In this case, we need exactly UTF-16
https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/76f10dd8-699d-45e6-a53c-5aefc586da20

Reply:
by way

#ifdef UNICODE
 typedef LPCWSTR LPCTSTR; 
#else
 typedef LPCSTR LPCTSTR;
#endif

#ifdef UNICODE
  #define SetFileAttributes SetFileAttributesW; 
#else
  #defineSetFileAttributes SetFileAttributesA;
#endif

Reply:
So `reinterpret_cast<LPCTSTR>(backupDirNativePath.utf16())` is the winner, I guess

Reply:
> and might even work independant of UNICODE being #define'd

Not quite. If `UNICODE` is not defined then `SetFileAttributes` and `LPCTSTR` expect an 8-bit string, not a 16-bit string like `QString::utf16()` provides. We could use `QString::toLocal8Bit().constData()` in that case, but there is a better way:

```c++
SetFileAttributesW(reinterpret_cast<LPCWSTR>(backupDirNativePath.utf16()), FILE_ATTRIBUTE_HIDDEN);
```

Notice explicit use of the `SetFileAttributesW` and `LPCWSTR` rather than the aliases `SetFileAttributes` and `LPCTSTR`, so this should work on modern Windows regardless of whether `UNICODE` is defined.

As I understand it, undefining `UNICODE` only changes the aliases; it doesn't actually disable Unicode support via the explicit `W` functions. The aliases are necessary for compatibility with code written prior to Windows NT, but I don't think we need to go there!

Reply:
But that other error still requires UNICODE, right? Or is there a method to make it waterproof too?

Reply:
yes, there we can also use the method with the `W` suffix

Reply:
There is just a choice, either uses Windows API methods with the `W` suffix everywhere or define the default `UNICODE` by default.

Reply:
Also for the warning generated a few lines later?

Reply:
I've updated the PR to use the `W` suffix in both files. The warning is different and I don't have time to investigate further unfortunately. I recommend merging this as-is to fix the error for MinGW users. The warning can wait for another time.